### PR TITLE
Add files for Docker

### PR DESCRIPTION
**Pull request by BMOD89**
Add files to build and start the application in a Docker container. 
I needed to update the dependencies to be able to build the server application. 
I also found a typo in one of the links.

### CODE REVIEW

**Comment by BMOD89** [comment]
@abrain I applied your suggestions from https://github.com/alarmdisplay/display/pull/3 in this repository as well apart from using node:lts. 

I have compared the versions of tesseract between node:lts (debian stretch) node:lts-buster and ubuntu:20.04. The versions are 3.04.01, 4.0.0 and 4.1.1. I am not sure if this makes any difference. Should I also use another base image?

**Comment by abrain** [comment]
> I have compared the versions of tesseract between node:lts (debian stretch) node:lts-buster and ubuntu:20.04. The versions are 3.04.01, 4.0.0 and 4.1.1. I am not sure if this makes any difference. Should I also use another base image?

We'll want version 4 for sure. In this case I'd also opt for the Ubuntu base image as its packages seem to be updated more frequently.

**Comment by BMOD89** [comment]
All right. The docker changes in this PR should have a similar quality as in the display PR. 
If you have further questions, let me know

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+RUN npm install --no-audit && npm run build
+
+FROM prebuilder as build-server
+
+COPY ./server /home/node/app
+WORKDIR /home/node/app
+
+RUN npm install --no-audit && npm run compile
+
+FROM prebuilder
```
Comment (abrain) [inline Dockerfile:29]: As we are not using the `node` base image here, we have to create the non-privileged user ourselves.
See https://github.com/nodejs/docker-node/blob/0544a5176c2a92c58c0809f5082e2877d796d394/14/buster/Dockerfile#L3-L4

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+COPY ./server/package*.json /home/node/app/
+COPY --from=build-console /usr/local/hub/console/dist /home/node/app/ext-console
+
+WORKDIR /home/node/app
+RUN npm install --only=production --no-audit
+
+EXPOSE 3030
+ENV NODE_ENV production
+ENV NODE_CONFIG_ENV docker
+CMD [ "node", "index.js" ]
```
Comment (abrain) [inline Dockerfile:42]: The container should be run as the `node` user

File: .dockerignore [hunk 3]
```diff
@@ -0,0 +1 @@
+server/node_modules/
```
Comment (abrain) [inline .dockerignore:1]:
```suggestion
**/node_modules/
server/config/local*.json
```
The `console` folder also has `node_modules` which should be ignored. The configuration mechanism for the server is the same as for the Display server, so the local files should also be ignored.

**Review by abrain — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (BMOD89) [reply Dockerfile:46]: done

File: Dockerfile (hunk 1 above)
Comment (BMOD89) [reply Dockerfile:29]: done

**Comment by BMOD89** [comment]
I am currently struggling with the user rights and the text recognition when I do not use the root user. I need to solve this first. Therefor I set the PR to draft

**Comment by BMOD89** [comment]
I have updated the files. I need to create the uploads folder in /home/node/app with the option -m777 so that the app has access rights to that folder. The other files belong to the root user. I would prefer that because of the write protection. @abrain what do you think?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+RUN apt-get update && apt-get install -y \
+    git \
+    imagemagick \
+    nodejs \
+    npm \
+    tesseract-ocr-deu \
+    && rm -rf /var/lib/apt/lists/*
+
+# allow ImageMagick to process PDFs
+RUN sed -i '$i\ \ <policy domain="coder" rights="read | write" pattern="PDF" />' /etc/ImageMagick-6/policy.xml
```
[the remark this answers is not in the document]
Comment (BMOD89) [reply Dockerfile]: done

**Comment by abrain** [comment]
So, in the meantime I played around with the build process of the Display component. There's now a build script, that puts everything together, ready to use or package. The Dockerfile also had this build process in it, and it felt considerably slower to build it in Docker. So I shortened the Dockerfile to use the result from the build script (see https://github.com/alarmdisplay/display/compare/1f1ad40a910eeff32adb35ae1f37b562bad57756...3e02a35f12cd9ee13ac1125ece54fe115b5b93a9).

I added such a build script to the hub repo as well and would suggest to use the same approach. Or is there a specific reason to build it in Docker, that I am overlooking here?

**Comment by BMOD89** [comment]
I my opinion this is fine if you develop on your local machine and have every tools installed. 
In my case, i don't have all the node.js tools on my local machine. I relied on building all the stuff inside the docker container. So with the current Dockerfile I wouldn't be able to build the container anymore from scratch. 
(Even for debugging I have used a container and the "Remote - Containers" extension of VS Code :smile:)

I agree that building the container is very slow. One of the main reasons is that the dependencies need to be downloaded if one of the source files changed. 
In the meantime I got some inspirations on how to speed up the build process. Do you want to keep it like it is or should i give it a try?

**Comment by BMOD89** [comment]
I have added the improved Dockerfile in that branch https://github.com/BMOD89/hub/tree/feature/speedup_docker_build. It should avoid downloading the dependencies for every change. I have also changed the Dockerfile in my display fork.  

This page describes how to further improve the build time by using buildkit https://geshan.com.np/blog/2020/10/docker-build-example-faster-docker-build/

**Comment by abrain** [comment]
Ok, that sounds reasonable. The Docker-driven development style isn't for me (at least for this project 😄), but I see your point. Also the caching mechanism with copying the npm files first looks promising. I'm fine with making the main Dockerfile the full multi-stage build and then possibly have a second Dockerfile for building the image from the prebuilt application.

I'm using BuildKit already for build the image for multiple architectures 👍🏻 Let's give the speed-up version a try.

Btw, we can save another layer by inlining the environment variable for `apt-get` like so: `RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \ ...`

Also, I got rid of the uploads service while reworking the OCR and text analysis flow. So the uploads folder / volume is no longer necessary.

I've also reverted the shortening of the Dockerfile in the Display repository, feel free to submit a PR for the speed-up version there as well.

**Comment by BMOD89** [comment]
I have rebased my changes to the latest develop branch. 
@abrain should I remove the Readme commit? 
What do you think about the non-Dockerfile changes?

**Comment by abrain** [comment]
> @abrain should I remove the Readme commit?
> What do you think about the non-Dockerfile changes?

I think, the READMEs will have to be rewritten anyways but I'd leave the PR as it is now.

Thanks for the work and bearing with me and my change requests ;)

**Review by abrain — APPROVED** [review]

**Pull request merged**